BitwardenShared/Core/Vault/Services/TOTP/OTPAuthModel.swift

```diff
@@ -4,49 +4,101 @@ import Foundation
+    let algorithm: TOTPCryptoHashAlgorithm
+
+    /// The number of digits in the OTP.
+    let digits: Int
+
+    /// The provider or service the account is associated with.
+    let issuer: String?
+
     /// The Base32-encoded key used for generating the OTP.
     let keyB32: String
 
     /// The time period in seconds for which the OTP is valid.
     let period: Int
 
-    /// The number of digits in the OTP.
-    let digits: Int
+    /// The unparsed key URI.
+    let uri: String
 
-    /// The hashing algorithm used for generating the OTP.
-    let algorithm: TOTPCryptoHashAlgorithm
+    // MARK: Initialization
 
     /// Initializes a new instance of `OTPAuthModel`.
     ///
     /// - Parameters:
+    ///   - account: The name of the account for the key.
+    ///   - algorithm: The hashing algorithm used for generating the OTP.
+    ///   - digits: The number of digits in the OTP.
+    ///   - issuer: The provider or service the account is associated with.
     ///   - keyB32: The Base32-encoded key.
     ///   - period: The time period in seconds for which the OTP is valid.
-    ///   - digits: The number of digits in the OTP.
-    ///   - algorithm: The hashing algorithm used for generating the OTP.
-    init(keyB32: String, period: Int, digits: Int, algorithm: TOTPCryptoHashAlgorithm) {
+    ///   - uri: The unparsed key URI.
+    ///
+    init(
+        accountName: String? = nil,
+        algorithm: TOTPCryptoHashAlgorithm,
+        digits: Int,
+        issuer: String? = nil,
+        keyB32: String,
+        period: Int,
+        uri: String
+    ) {
+        self.accountName = accountName
+        self.algorithm = algorithm
+        self.digits = digits
+        self.issuer = issuer
         self.keyB32 = keyB32
         self.period = period
-        self.digits = digits
-        self.algorithm = algorithm
+        self.uri = uri
     }
 
     /// Parses an OTP Auth URI into its components.
     ///
     /// - Parameter otpAuthKey: A string representing the OTP Auth URI.
     ///
     init?(otpAuthKey: String) {
-        guard let urlComponents = URLComponents(string: otpAuthKey.lowercased()),
+        guard let urlComponents = URLComponents(string: otpAuthKey),
               urlComponents.scheme == "otpauth",
               let queryItems = urlComponents.queryItems,
               let secret = queryItems.first(where: { $0.name == "secret" })?.value,
               secret.uppercased().isBase32 else {
             return nil
         }
 
+        var accountName: String?
+        var issuer: String?
+        if let label = urlComponents.url?.lastPathComponent {
+            let parts = label.split(separator: ":")
+            if parts.count > 1 {
+                issuer = String(parts[0])
+                accountName = String(parts[1])
+            } else {
+                accountName = label
+            }
+        }
+
+        if issuer == nil {
+            issuer = queryItems.first { $0.name == "issuer" }?.value
+        }
```

Comment:
🤔 The OTP specs—particularly Google's—are maddeningly inconsistent and ambiguous about how to handle things like the issuer. It seemed to me like Google had regretted putting Issuer in the label, and therefore moved it to being another query parameter, but had to keep it around for backwards compatibility.

In [the Authenticator app](https://github.com/bitwarden/authenticator-ios/blob/main/AuthenticatorShared/Core/Vault/Services/TOTP/OTPAuthModel.swift#L101) I went the other way, treating the parameter as canonical and then stripping it out of the label if it was there to get the account name. Basically following this guidance from [Google's spec](https://github.com/google/google-authenticator/wiki/Key-Uri-Format):

> If the issuer parameter is absent, issuer information may be taken from the issuer prefix of the label.

I'm not sure if we want both apps to actually have the same implementation; in practice the two are I think functionally equivalent, it mostly comes down to how they handle incorrectly URIs that aren't to spec.

Reply:
Ah, that makes sense to me, I can swap those around.